src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -0,0 +1,218 @@
+
+                // Create driver that holds onto compilation and associated analyzers
+                return compilation.WithAnalyzers(analyzers, analysisOptions);
+            }
+
+            private CompilationWithAnalyzersOptions GetAnalyzerOptions(
+                Project project,
+                bool logAnalyzerExecutionTime,
+                bool reportSuppressedDiagnostics)
+            {
+                // in IDE, we always set concurrentAnalysis == false otherwise, we can get into thread starvation due to
+                // async being used with syncronous blocking concurrency.
+                return new CompilationWithAnalyzersOptions(
+                    options: new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Workspace),
+                    onAnalyzerException: GetOnAnalyzerException(project.Id),
+                    analyzerExceptionFilter: GetAnalyzerExceptionFilter(project),
+                    concurrentAnalysis: false,
+                    logAnalyzerExecutionTime: logAnalyzerExecutionTime,
+                    reportSuppressedDiagnostics: reportSuppressedDiagnostics);
+            }
+
+            private Func<Exception, bool> GetAnalyzerExceptionFilter(Project project)
+            {
+                return ex =>
+                {
+                    if (project.Solution.Workspace.Options.GetOption(InternalDiagnosticsOptions.CrashOnAnalyzerException))
+                    {
+                        // if option is on, crash the host to get crash dump.
+                        FatalError.ReportUnlessCanceled(ex);
+                    }
+
+                    return true;
+                };
+            }
+
+            private Action<Exception, DiagnosticAnalyzer, Diagnostic> GetOnAnalyzerException(ProjectId projectId)
+            {
+                return _owner.Owner.GetOnAnalyzerException(projectId, _owner.DiagnosticLogAggregator);
+            }
+
+            private void ResetAnalyzerDriverMap()
+            {
+                // we basically eagarly clear the cache on some known changes
+                // to let CompilationWithAnalyzer go.
+
+                // we create new conditional weak table every time, it turns out 
+                // only way to clear ConditionalWeakTable is re-creating it.
+                // also, conditional weak table has a leak - https://github.com/dotnet/coreclr/issues/665
+                _map = new ConditionalWeakTable<Project, CompilationWithAnalyzers>();
+            }
+
+            [Conditional("DEBUG")]
+            private void AssertAnalyzers(CompilationWithAnalyzers analyzerDriver, IEnumerable<StateSet> stateSets)
+            {
+                if (analyzerDriver == null)
+                {
+                    // this can happen if project doesn't support compilation or no stateSets are given.
+                    return;
+                }
+
+                // make sure analyzers are same.
```

Comment:
I recently changed this assert in v1 to handle suppressed analyzers - if you change address my comment above, you should even change this code to do suppressed analyzer filtering